src/main/java/org/folio/search/utils/CallNumberUtils.java

```diff
@@ -144,6 +153,75 @@ public static Long getCallNumberAsLong(String callNumber, int firstPosition) {
     return callNumberToLong(callNumber, startVal, CN_MAX_CHARS - 1);
   }
 
+  /**
+   * Excludes irrelevant items from result.
+   *
+   * <p>
+   * This algorithm takes call number browse result and call number type and removes irrelevant items from result
+   * approach:
+   * <ul>
+   *   <li>Each call number browse item has its own fullCallNumber field, which may vary from its instance's item or
+   *   holding call number. Matching ones filtered </li>
+   *   <li>CallNumberBrowseItem's instance may have items or holdings
+   *   which may have call numbers with different type. They will also be filtered by callNumberType</li>
+   *   <li>all filtered records are added together and returned</li>
+   * </ul>
+   * </p>
+   *
+   * @param callNumberType - call number type to check/compare result items' types
+   * @param records - list of CallNumberBrowseItem objects
+   * @return filtered records
+   */
+  public static List<CallNumberBrowseItem> excludeIrrelevantResultItems(String callNumberType,
+                                                                        List<CallNumberBrowseItem> records) {
+    if (StringUtils.isBlank(callNumberType) || records == null || records.isEmpty()) {
+      return records;
+    }
+
+    records.forEach(r -> {
+      r.getInstance().setItems(getItemsFiltered(callNumberType, r));
+      r.getInstance().setHoldings(getHoldingsFiltered(callNumberType, r));
+    });
+    return records;
+  }
+
+  @NotNull
+  private static List<@Valid Holding> getHoldingsFiltered(String callNumberType, CallNumberBrowseItem item) {
+    return item.getInstance().getHoldings()
+      .stream()
+      .filter(h -> isInGivenType(callNumberType, h.getCallNumber()))
```

Comment:
Why don't you check it by callNumberTypeId like for items?

Reply:
Why do we need to filter both holdings and items? Do we get call numbers from both holdings and items while converting elastic search responses?

Reply:
1. Because in holdings there is no type information(id or name of type).
2. I am filtering both items and holdings call numbers because in the ticket it is being claimed that irrelevant call numbers from holdings are coming up while browsing.